StringName Dictionary keys

rune-scape opened the pull request:
follow up for https://github.com/godotengine/godot/pull/68747
removes conversion from StringName to String in Dictionary

it seems like a simple change, but many places still compare a Variants type with STRING
simply removing the conversion gave me alot of errors, so i did the sane thing and looked at every instance of an equality comparison to `Variant::STRING`, and updated most of them with an `is_string()` method i added to variant:
-- if it accepted String and StringName, i updated it to use `is_string()` (might as well, it's much easier to read)
-- if it was a closed system, (as in no external input, the type is always known to be string), i let it be
-- if it accepted input, (especially from GDScript), i updated it
-- if i couldn't tell for sure if it was a closed system (like in `editor_debugger_inspector.cpp`):
---- if it gave an error for StringName input, i left it (could be caught later)
---- if it silently failed, i updated it

note: Variant does not convert StringName to Color or to NodePath, i'm not sure whether it makes sense to accept StringName, so i let it be for now

fixes https://github.com/godotengine/godot/issues/68834

here are some numbers i got from this example, with different length keys (on a production build):
[TestDictIndexing.zip](https://github.com/godotengine/godot/files/10241791/TestDictIndexing.zip):
```
"%s":
  2.318s: String keys create
 81.679s: StringName keys create
  7.384s: String assign create
  1.973s: String dict String index read
  2.145s: String dict StringName index read
  7.278s: StringName assign create
  2.324s: StringName dict String index read
  1.788s: StringName dict StringName index read

"longer-test-key-%s":
  3.883s: String keys create
 83.304s: StringName keys create
  8.323s: String assign create
  2.406s: String dict String index read
  2.462s: String dict StringName index read
  7.543s: StringName assign create
  2.792s: StringName dict String index read
  1.732s: StringName dict StringName index read
```
with a long enough test key, StringName is faster even when finding a String key:
```
"how-long-could-it-take-to-hash-this-string?-key-%s":
  4.566s: String keys create
 82.376s: StringName keys create
 10.346s: String assign create
  2.949s: String dict String index read
  2.697s: String dict StringName index read
  8.529s: StringName assign create
  3.408s: StringName dict String index read
  1.727s: StringName dict StringName index read
```

KoBeWi commented:
I tested this time and it fixes #68834 for real, without affecting usability.

rune-scape committed:
StringName Dictionary keys

also added 'is_string()' method to Variant
and refactored many String type comparisons to use it instead

Repiteo reviewed (approved):
Nice performance boost & helps iron out those last few string/stringname discrepancies, LGTM!

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/aaa4560729bf0161deb71789b47eba5623893845), it works as expected.

However, I can't really discern a noticeable performance difference before and after this PR. It's slightly faster in some cases (`StringName assign create`) but is slower in others (`StringName dict String index read`).

Using a release x86_64 Linux export template binary (`optimize=speed lto=full`) for running the MRP.

Binary size with this PR is 4 KB smaller compared to `master`.

### Before

```
  1.451s: String keys create
 33.451s: StringName keys create
  4.333s: String assign create
  0.482s: String dict String index read
  0.546s: String dict StringName index read
  4.511s: StringName assign create
  0.503s: StringName dict String index read
  0.561s: StringName dict StringName index read
```

### After

```
  1.473s: String keys create
 33.515s: StringName keys create
  4.403s: String assign create
  0.483s: String dict String index read
  0.566s: String dict StringName index read
  4.322s: StringName assign create
  0.638s: StringName dict String index read
  0.461s: StringName dict StringName index read
```

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 39)
</details>

rune-scape commented:
it may be because StringName == String doesnt do an in-place comparison like it could, copying the internal String of the StringName (or allocating a new String if not from GDScript) before comparing it leading to an extra ref then deref, ive already made a fix including this change on my branch, ill find some time to test production builds with the settings u mentioned on all 3 variants [before PR, after PR, after PR with performance fix]

rune-scape commented:
i changed the test script to autorun and quit and spend less time making keys and more time on the actual test: 
[TestDictIndexing2.zip](https://github.com/godotengine/godot/files/15493077/TestDictIndexing2.zip)

table of percent time of execution compared to before this PR:
![image](https://github.com/godotengine/godot/assets/55224592/d46d0344-8482-4022-8af4-bfd27982471d)

`String keys create`/`StringName keys create` of the large dict test are maybe good anchors for how inconsistent the results are: ~ +/-2%
theres clearly a big slowdown after this PR in `StringName_dict[String]` thats solved by in-place comparisons


theres also the 101.4% `StringName_dict[String]` that 'should' be the exact same or slightly less but thats within the +/-2%

command `scons target=template_release optimise=speed lto=full scu_build=yes`
branch: https://github.com/rune-scape/godot/tree/stringname-dict2

<details><summary>outputs</summary>

- before:
<pre><code>Godot Engine v4.3.beta.custom_build.4bb8c06f6 (2024-05-15 20:42:02 UTC) - https://godotengine.org

key_count/dict_size: 1000000
iterations: 20
  0.47406s: String keys create
  0.41555s: StringName keys create
 12.63348s: String_dict create
 13.42629s: StringName_dict create
  6.92510s: String_dict[String]
  7.23151s: String_dict[StringName]
  7.30472s: StringName_dict[StringName]
  6.96156s: StringName_dict[String]
key_count/dict_size: 50
iterations: 1000000
  0.00003s: String keys create
  0.00002s: StringName keys create
  9.35630s: String_dict create
 11.17116s: StringName_dict create
  4.94667s: String_dict[String]
  6.02697s: String_dict[StringName]
  6.15707s: StringName_dict[StringName]
  4.91070s: StringName_dict[String]</code></pre>
- after:
<pre><code>Godot Engine v4.3.beta.custom_build.fc60dbe4f (2024-05-16 00:01:37 UTC) - https://godotengine.org

key_count/dict_size: 1000000
iterations: 20
  0.47541s: String keys create
  0.39441s: StringName keys create
 12.59615s: String_dict create
 12.29998s: StringName_dict create
  6.98052s: String_dict[String]
  7.20852s: String_dict[StringName]
  6.70885s: StringName_dict[StringName]
  7.46362s: StringName_dict[String]
key_count/dict_size: 50
iterations: 1000000
  0.00003s: String keys create
  0.00002s: StringName keys create
  9.36504s: String_dict create
  9.08053s: StringName_dict create
  4.97574s: String_dict[String]
  6.04046s: String_dict[StringName]
  4.82189s: StringName_dict[StringName]
  6.00519s: StringName_dict[String]</code></pre>
- after w/ in-place comparison:
<pre><code>Godot Engine v4.3.beta.custom_build.5a312e76e (2024-05-28 11:15:25 UTC) - https://godotengine.org

key_count/dict_size: 1000000
iterations: 20
  0.47619s: String keys create
  0.40128s: StringName keys create
 12.61830s: String_dict create
 12.50513s: StringName_dict create
  6.93993s: String_dict[String]
  6.80969s: String_dict[StringName]
  6.80732s: StringName_dict[StringName]
  6.99365s: StringName_dict[String]
key_count/dict_size: 50
iterations: 1000000
  0.00003s: String keys create
  0.00002s: StringName keys create
  9.38292s: String_dict create
  9.08908s: StringName_dict create
  4.93871s: String_dict[String]
  5.04735s: String_dict[StringName]
  4.74848s: StringName_dict[StringName]
  4.97938s: StringName_dict[String]</code></pre>
</details>
<details><summary>perf stats (idk if this is even useful)</summary>

- before:
<pre><code># started on Wed May 29 09:19:01 2024


 Performance counter stats for './bin/godot.linuxbsd.template_release.x86_64.stringname-dict2-0-g4bb8c06f6a --headless ../../Downloads/TestDictIndexing/project.godot':

          99111.20 msec task-clock                       #    1.000 CPUs utilized             
             13853      context-switches                 #  139.772 /sec                      
                72      cpu-migrations                   #    0.726 /sec                      
             75056      page-faults                      #  757.291 /sec                      
      304866650326      cycles                           #    3.076 GHz                       
      577664589537      instructions                     #    1.89  insn per cycle            
      101930137819      branches                         #    1.028 G/sec                     
         179113228      branch-misses                    #    0.18% of all branches           

      99.068539644 seconds time elapsed

      98.726199000 seconds user
       0.313073000 seconds sys
</code></pre>
- after:
<pre><code># started on Wed May 29 09:25:25 2024


 Performance counter stats for './bin/godot.linuxbsd.template_release.x86_64.stringname-dict2-0-gfc60dbe4f0 --headless ../../Downloads/TestDictIndexing/project.godot':

          95753.77 msec task-clock                       #    0.999 CPUs utilized             
             13376      context-switches                 #  139.692 /sec                      
                23      cpu-migrations                   #    0.240 /sec                      
             74266      page-faults                      #  775.593 /sec                      
      294240378895      cycles                           #    3.073 GHz                       
      554352902206      instructions                     #    1.88  insn per cycle            
       96495768265      branches                         #    1.008 G/sec                     
         180022020      branch-misses                    #    0.19% of all branches           

      95.861765139 seconds time elapsed

      95.313911000 seconds user
       0.373233000 seconds sys
</code></pre>
- after w/ in-place comparison:
<pre><code># started on Wed May 29 09:28:41 2024


 Performance counter stats for './bin/godot.linuxbsd.template_release.x86_64.stringname-dict2-0-g5a312e76e4 --headless ../../Downloads/TestDictIndexing/project.godot':

          93036.65 msec task-clock                       #    0.998 CPUs utilized             
             13207      context-switches                 #  141.955 /sec                      
                82      cpu-migrations                   #    0.881 /sec                      
             75291      page-faults                      #  809.262 /sec                      
      286030673926      cycles                           #    3.074 GHz                       
      544972454573      instructions                     #    1.91  insn per cycle            
       94114936119      branches                         #    1.012 G/sec                     
         171548871      branch-misses                    #    0.18% of all branches           

      93.184708944 seconds time elapsed

      92.546298000 seconds user
       0.422159000 seconds sys
</code></pre>
</p>
</details>

KoBeWi commented:
Identifier access (i.e. Lua-style) will always use StringNames. Making it use String would force a coversion.

dalexeev commented:
You are probably right from a performance perspective. I meant compatibility and consistency issues. In 4.3 there is no difference between Lua- and Python-style dictionaries:

```gdscript
var dict1 := { x = 1 }
var dict2 := { "x": 1 }
print(dict1) # { "x": 1 }
print(dict2) # { "x": 1 }
```

KoBeWi commented:
The performance impact is what #68834 is about.

dalexeev reviewed (approved):
Makes sense to me, and is essential for supporting typed dictionaries. It has some quirks, but the tradeoff between performance and compatibility seems worth it. `String` and `StringName` are interchangeable in most cases, so this shouldn't cause major compatibility breaks. However, users who explicitly check the key type or rely on its string representation may have some issues.

rune-scape commented:
rebased. and looked through comparisons to `Variant::STRING` again and changed a few more

also i recommend that https://github.com/godotengine/godot/pull/92570 gets merged too, so the performance gets better in all cases and not worse in some

akien-mga commented:
```gdscript
var dict1 := { x = 1 }
var dict2 := { "x": 1 }
print(dict1) # { &"x": 1 }
print(dict2) # { "x": 1 }
```

I'm quite concerned about this, I can really see users running into unexpected bugs when `dict1.x == dict2.x` will be false, and they won't understand why.

I'm not sure it's worth it to introduce this kind of inconsistency in the API just based on how you declared a dictionary, for a relatively minor performance gain. The tradeoff seems bad _a priori_ (didn't review the changes and discussion in depth).

KoBeWi commented:
> I can really see users running into unexpected bugs when dict1.x == dict2.x will be false, and they won't understand why.

From what I understand, this would return `true` actually. The idea is that Dictionaries behave the same as currently, but internally allow StringNames.

rune-scape commented:
yes dictionaries should behave the same as before (because of StringLikeVariantComparator), just that retrieving a key might give you a string name when it gave a string before (meaning comparing the type of a key to `TYPE_STRING` might not work anymore, maybe there should be an `is_string` global gdscript method for convenience and to make updating code easier)

theres a later PR i have to add support for iterating stringname characters with a for loop (relies on https://github.com/godotengine/godot/pull/92570)
might be good to have, with another way of leaking more stringnames into code

theres also less than and greater than comparisons but to match that to string, the behavior of the stringname comparisons would need to change, so probly not

vnen commented:
> 1. The following behavior may be confusing:
> 
> ```gdscript
> @tool
> extends EditorScript
> 
> func _run() -> void:
>     var dict := {
>         #"x": 1,
>         #&"x": 2, # Error: Key "x" was already used in this dictionary (at line 6).
>     }
>     dict["x"] = 1
>     dict[&"x"] = 2
>     dict[&"y"] = 3
>     dict["y"] = 4
>     print(dict) # { "x": 2, &"y": 4 }
> ```

This is the same as the current behavior, the only difference is that currently all keys will be strings even if you us StringName. While in this case it might feel odd, I believe it is expected that using String to access a StringName key just replaces the value, not the key (and in practice does not even matter much, given how interchangeable they are).


> 2. There are some compatibility breakage and inconsistency between Lua- and Python-style dictionaries:
> 
> ```gdscript
> var dict1 := { x = 1 }
> var dict2 := { "x": 1 }
> print(dict1) # { &"x": 1 }
> print(dict2) # { "x": 1 }
> ```

The only breakage is printing pretty much. I don't think we guarantee compatibility to that level, relying on printed serialization was always finicky.

> 1. As far as I remember, `{ &"x": 1 } == { "x": 1 }`. Please correct me if I am wrong.

That is correct. Even with this PR this is still true.

----

IMO the trade-off is good. The Lua-style was always supposed to use StringNames and only changed because of the issues with the type difference. But those issues are pretty much solved now, so it makes sense to allow users have better performance when they need it.

vnen commented on modules/gdscript/tests/scripts/runtime/features/dictionary_string_stringname_equivalent.gd:
```diff
@@ -7,7 +7,7 @@ func test():
+	print("StringName key is TYPE_STRING_NAME: ", typeof(stringname_dict.keys()[0]) == TYPE_STRING_NAME)
 
 	print("StringName gets String: ", string_dict.get(&"abc"))
 	print("String gets StringName: ", stringname_dict.get("abc"))
```
There's a comment below this that is not true anymore:
`# They compare equal because StringName keys are converted to String.`

It should be changed or removed.

vnen reviewed (changes requested):
Small nitpick, otherwise looks good to me.

vnen reviewed (approved):
LGTM

akien-mga commented:
Thanks!

Faless commented:
> This is the same as the current behavior, the only difference is that currently all keys will be strings even if you us StringName.

One side effect of this change is that the following code:
```gdscript
extends Node

func _ready() -> void:
	var dict := {
		&"B": "test",
		&"A": "test"
	}
	var keys := dict.keys()
	keys.sort() # Will sort strings alphabetically, but StringNames by pointer.
	print(keys)
```

Prints:

```
["A", "B"] # Until Godot 4.3
[&"B", &"A"] # Since Godot 4.4 (after this PR)
```

While this is a consequence of the StringName not being casted to String, the resulting behavior may not be so obvious.

Since I suspect this case might be quite common across people updating, I would suggest we mention it (prominently) in the release notes.
